frame/staking/src/lib.rs

```diff
@@ -1320,113 +1602,76 @@ impl<T: Trait> Module<T> {
 		}
 	}
 
-	/// Reward a given validator by a specific amount. Add the reward to the validator's, and its
-	/// nominators' balance, pro-rata based on their exposure, after having removed the validator's
-	/// pre-payout cut.
-	fn reward_validator(stash: &T::AccountId, reward: BalanceOf<T>) -> PositiveImbalanceOf<T> {
-		let off_the_table = Self::validators(stash).commission * reward;
-		let reward = reward.saturating_sub(off_the_table);
-		let mut imbalance = <PositiveImbalanceOf<T>>::zero();
-		let validator_cut = if reward.is_zero() {
-			Zero::zero()
-		} else {
-			let exposure = Self::stakers(stash);
-			let total = exposure.total.max(One::one());
-
-			for i in &exposure.others {
-				let per_u64 = Perbill::from_rational_approximation(i.value, total);
-				imbalance.maybe_subsume(Self::make_payout(&i.who, per_u64 * reward));
+	/// Plan a new session potentially trigger a new era.
+	fn new_session(session_index: SessionIndex) -> Option<Vec<T::AccountId>> {
+		if let Some(current_era) = Self::current_era() {
+			// Initial era has been set.
+
+			let current_era_start_session_index = Self::eras_start_session_index(current_era)
+				.unwrap_or_else(|| {
+					frame_support::print("Error: start_session_index must be set for current_era");
+					0
+				});
+
+			let era_length = session_index.checked_sub(current_era_start_session_index)
+				.unwrap_or(0); // Must never happen.
+
+			match ForceEra::get() {
+				Forcing::ForceNew => ForceEra::kill(),
+				Forcing::ForceAlways => (),
+				Forcing::NotForcing if era_length >= T::SessionsPerEra::get() => (),
+				_ => return None,
 			}
 
-			let per_u64 = Perbill::from_rational_approximation(exposure.own, total);
-			per_u64 * reward
-		};
-
-		imbalance.maybe_subsume(Self::make_payout(stash, validator_cut + off_the_table));
-
-		imbalance
+			Self::new_era(session_index)
+		} else {
+			// Set initial era
+			Self::new_era(session_index)
+		}
 	}
 
-	/// Session has just ended. Provide the validator set for the next session if it's an era-end.
-	fn new_session(session_index: SessionIndex) -> Option<Vec<T::AccountId>> {
-		let era_length = session_index.checked_sub(Self::current_era_start_session_index()).unwrap_or(0);
-		match ForceEra::get() {
-			Forcing::ForceNew => ForceEra::kill(),
-			Forcing::ForceAlways => (),
-			Forcing::NotForcing if era_length >= T::SessionsPerEra::get() => (),
-			_ => return None,
+	/// Start a session potentially starting an era.
+	fn start_session(start_session: SessionIndex) {
+		let next_active_era = Self::active_era().map(|e| e + 1).unwrap_or(0);
+		if let Some(next_active_era_start_session_index) = Self::eras_start_session_index(next_active_era) {
+			if next_active_era_start_session_index == start_session {
```

Comment:
```suggestion
			if next_active_era_start_session_index <= start_session {
```

We can safely make this a less than or equal to right?

Reply:
oh yes definitely
Also would be great to print a comment in case this `<` happen saying that something wrong happens, probably some sessions has been skipped.
(or at least but into comment that `<` should never happen but it is better in case some session has been skipped somehow. (or if two era start at the same session index..))

Reply:
thanks for having this done, looks good to me